lib/explorer/polars_backend/series.ex

```diff
@@ -343,18 +348,27 @@ defmodule Explorer.PolarsBackend.Series do
       |> List.duplicate(PolarsSeries.length(lhs))
       |> Series.from_list()
 
-  defp encode_date(%Date{} = date), do: Date.to_iso8601(date)
+  defp encode_date(%Date{} = date) do
+    Date.to_gregorian_days(date) - @unix_epoch_in_gregorian_days
+  end
+
   defp encode_date(date) when is_nil(date), do: nil
 
-  defp encode_datetime(%NaiveDateTime{} = datetime), do: NaiveDateTime.to_iso8601(datetime)
-  defp encode_datetime(%DateTime{} = datetime), do: DateTime.to_iso8601(datetime)
+  defp encode_datetime(%module{} = datetime) when module in [NaiveDateTime, DateTime] do
+    {secs, microsecs} = module.to_gregorian_seconds(datetime)
+
+    (secs - @unix_epoch_in_gregorian_secs) * 1_000 + div(microsecs, 1000)
+  end
+
   defp encode_datetime(datetime) when is_nil(datetime), do: nil
 
   defp decode_date(date) when is_integer(date), do: Date.add(~D[1970-01-01], date)
   defp decode_date(date) when is_nil(date), do: nil
 
-  defp decode_datetime(date) when is_integer(date),
-    do: NaiveDateTime.add(~N[1970-01-01 00:00:00], date, :millisecond)
+  # Note that we lost microseconds in the conversion, since Polars only works with milliseconds.
+  defp decode_datetime(epoch_date_in_ms) when is_integer(epoch_date_in_ms) do
+    DateTime.from_unix!(epoch_date_in_ms, :millisecond) |> DateTime.to_naive()
```

Comment:
I would expect the NaiveDateTime.add version (the previous one) to be more efficient, isn't that the case?

Reply:
@josevalim they are almost identical. The only difference is that the previous one didn't have the last part (the milliseconds). When I add that, then it gets slightly slower. Do you have a preference?

## Code under test

```elixir
today_in_days = Date.utc_today() |> Date.to_gregorian_days()
day_in_seconds = 86_400

random_datetime = fn size ->
  for _i <- 0..size do
    days = :rand.uniform(today_in_days)

    DateTime.from_gregorian_seconds(days * day_in_seconds + :rand.uniform(today_in_days))
    |> DateTime.to_unix(:millisecond)
  end |> List.first()
end

inputs = %{
  "rand-1" => random_datetime.(100),
  "rand-2" => random_datetime.(100),
  "rand-3" => random_datetime.(100),
}

Benchee.run(
  %{
    "naive add" => fn milli ->  NaiveDateTime.add(~N[1970-01-01 00:00:00.000], milli, :millisecond)end,
    "from_unix |> to naive" => fn milli -> DateTime.from_unix!(milli, :millisecond) |> DateTime.to_naive() end
  },
  inputs: inputs,
  time: 10,
  memory_time: 2
)
```

## Results

Operating System: Linux
CPU Information: AMD Ryzen 9 5950X 16-Core Processor
Number of Available Cores: 32
Available memory: 15.58 GB
Elixir 1.13.1
Erlang 24.1.5

Benchmark suite executing with the following configuration:
warmup: 2 s
time: 10 s
memory time: 2 s
parallel: 1
inputs: rand-1, rand-2, rand-3
Estimated total run time: 1.40 min

Benchmarking from_unix |> to naive with input rand-1...
Benchmarking from_unix |> to naive with input rand-2...
Benchmarking from_unix |> to naive with input rand-3...
Benchmarking naive add with input rand-1...
Benchmarking naive add with input rand-2...
Benchmarking naive add with input rand-3...

##### With input rand-1 #####
Name                            ips        average  deviation         median         99th %
from_unix |> to naive        2.61 M      383.12 ns  ±2659.97%         320 ns        1071 ns
naive add                    2.17 M      460.09 ns  ±4586.85%         400 ns        1091 ns

Comparison:
from_unix |> to naive        2.61 M
naive add                    2.17 M - 1.20x slower +76.98 ns

Memory usage statistics:

Name                     Memory usage
from_unix |> to naive           736 B
naive add                       576 B - 0.78x memory usage -160 B

**All measurements for memory usage were the same**

##### With input rand-2 #####
Name                            ips        average  deviation         median         99th %
from_unix |> to naive        2.57 M      389.30 ns   ±136.98%         360 ns         661 ns
naive add                    2.41 M      415.51 ns  ±4951.60%         359 ns        1071 ns

Comparison:
from_unix |> to naive        2.57 M
naive add                    2.41 M - 1.07x slower +26.21 ns

Memory usage statistics:

Name                     Memory usage
from_unix |> to naive           736 B
naive add                       576 B - 0.78x memory usage -160 B

**All measurements for memory usage were the same**

##### With input rand-3 #####
Name                            ips        average  deviation         median         99th %
from_unix |> to naive        2.46 M      406.64 ns   ±175.97%         380 ns         631 ns
naive add                    2.30 M      434.28 ns  ±1894.94%         380 ns        1091 ns

Comparison:
from_unix |> to naive        2.46 M
naive add                    2.30 M - 1.07x slower +27.64 ns

Memory usage statistics:

Name                     Memory usage
from_unix |> to naive           736 B
naive add                       576 B - 0.78x memory usage -160 B

**All measurements for memory usage were the same**

Reply:
That's surprising. :) The current way works for me then!